### PR TITLE
handover ADM file path to AtlasMap startup #7

### LINKED ISSUE #7: Open Atlasmap UI pre-filled when opening an "atlasmap transformation" file

### PR DESCRIPTION
**Pull request by lhein**
Signed-off-by: Lars Heinemann <lhein.smx@gmail.com>

This PR will make the Open AtlasMap an action on a ADM file in the explorer view. When executed it will pass the path of the adm file as `-Datlasmap.adm.path=<path>` to the startup of the atlasmap-standalone. 

For now I am not able to test if the file gets imported in AtlasMap correctly because https://github.com/atlasmap/atlasmap/issues/755 is still not implemented.

@pleacu @igarashitm FYI

fixes #7

### CODE REVIEW

File: src/extension.ts [hunk 1]
```diff
@@ -17,11 +17,15 @@ export function activate(context: vscode.ExtensionContext) {
 
 	let atlasmapExecutablePath = context.asAbsolutePath(path.join('jars','atlasmap-standalone.jar'));
 
-	context.subscriptions.push(vscode.commands.registerCommand('atlasmap.start', () => {
+	context.subscriptions.push(vscode.commands.registerCommand('atlasmap.start', (ctx) => {
+		let admFilePath: string;
+		if (ctx && ctx.path) {
+			admFilePath = ctx.path;
+		}
 		if (atlasMapLaunchPort === undefined) {
```
Comment (apupier) [inline src/extension.ts]: if I understand well, if there is already an AtlasMap UI started, the adm file won't be opened. We will just point to the existing AtlasMap Server instance which can be unrelated.

Comment (lhein) [reply src/extension.ts]: yes, valid point. We might enforce a stop in future to make sure that the adm file is processed. Will do that next.

Comment (lhein) [reply src/extension.ts]: added the force shutdown logic

**Comment by lhein** [comment]
> I think a test needs to be written (i know it can be tested right now but it is how TDD works in theory :-)) it will be an important use case with some possible corner cases.

When thinking more about it then it might be quite complicated to check if the import to AtlasMap was successful. We might need to check the HTML for some clue if the import worked. Not clear yet how to do it in a general way. From a logical point of view I would think that this needs to be tested on AtlasMap side. The only thing we can test is if the parameter is set correctly when spawning the process.

**Comment by apupier** [comment]
> > I think a test needs to be written (i know it can be tested right now but it is how TDD works in theory :-)) it will be an important use case with some possible corner cases.
> 
> When thinking more about it then it might be quite complicated to check if the import to AtlasMap was successful. We might need to check the HTML for some clue if the import worked. Not clear yet how to do it in a general way.

I have not said that it would be easy :-)
I expect that when using the webview, we would be able to access the content of the loaded DOM. Another way might be to request the available mappings through an API call to AtlasMap.

> From a logical point of view I would think that this needs to be tested on AtlasMap side. The only thing we can test is if the parameter is set correctly when spawning the process.

How to check that the parameter is set correctly? In case , there is a typo in it or the name is changing on AtlasMap side, we won't be able to detect it.
Furthermore, for now there is no test for this use case.

**Comment by lhein** [comment]
> How to check that the parameter is set correctly? In case , there is a typo in it or the name is changing on AtlasMap side, we won't be able to detect it.

The same applies if the API calls you mentioned change or the returned DOM model. There is always a risk.

**Comment by apupier** [comment]
> > How to check that the parameter is set correctly? In case , there is a typo in it or the name is changing on AtlasMap side, we won't be able to detect it.
> 
> The same applies if the API calls you mentioned change or the returned DOM model. There is always a risk.

I think that it is not the same. If the returned DOM model or the API changed, the test will be broken and we will notice that there is potentially something to change in the runtime product.
If we are only checking that the parameter is set when launched, the test will still pass even if the parameter is more correct.

**Comment by lhein** [comment]
valid point.

**Comment by lhein** [comment]
I will leave the improved tests for a separate PR (see #77)

The current state allows to open the AtlasMap view for a given ADM file by invoking the context menu "Open in AtlasMap" directly on the file. The webpage loaded the file fine in my local testing.

**Comment by lhein** [comment]
> I was expecting that it will open AtlasMap when double-clicking on the *.adm file

I am not sure if that is a good idea as we are not yet sure if there will probably be an edit option inside VSCode. So tbh I would maybe leave the dbl click for that instead.

**Comment by lhein** [comment]
> it is not working for me, I got a notification error "Failed to resolve the AtlasMap web UI." I have no error in logs.

This error occurs if it fails for more than 30 secs to receive the startup notification from the console output of AtlasMap afair.

File: src/extension.ts [hunk 2]
```diff
@@ -72,7 +102,12 @@ function launchAtlasMapLocally(atlasmapExecutablePath: string, port: string): Pr
 		requirements.resolveRequirements()
 			.then(requirements => {
 				let javaExecutablePath = path.resolve(requirements.java_home + '/bin/java');
-				atlasMapProcess = child_process.spawn(javaExecutablePath, ['-jar', atlasmapExecutablePath]);
+
+				if (admFilePath !== "") {										
+					atlasMapProcess = child_process.spawn(javaExecutablePath, ['-Datlasmap.adm.path=' + admFilePath, '-jar', atlasmapExecutablePath]);
```
[the remark this answers is not in the document]
Comment (lhein) [reply src/extension.ts]: thats probably the reason why it fails for you.

[the remark this answers is not in the document]
Comment (apupier) [reply src/extension.ts]: no, I checked that i don't have a space

**Comment by apupier** [comment]
> > it is not working for me, I got a notification error "Failed to resolve the AtlasMap web UI." I have no error in logs.
> 
> This error occurs if it fails for more than 30 secs to receive the startup notification from the console output of AtlasMap afair.

AtlasMap cannot start because the admFilePath provided is starting with a "/"
image


I tried to remove it in debug mode and it can start.

File: src/extension.ts [hunk 3]
```diff
@@ -103,7 +103,7 @@ function launchAtlasMapLocally(atlasmapExecutablePath: string, port: string, adm
 			.then(requirements => {
 				let javaExecutablePath = path.resolve(requirements.java_home + '/bin/java');
 
-				if (admFilePath !== "") {
+				if (admFilePath !== "") {
```
[the remark this answers is not in the document]
Comment (lhein) [reply src/extension.ts]: fixed

File: src/extension.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (lhein) [reply src/extension.ts]: seems to be not required.

**Comment by lhein** [comment]
Yeah, that is a problem when doing it so "unconnected". We have the requirement to always open the atlasmap ui with the specified file. That means I have to close a possibly running instance of AtlasMap as it could contain a different adm import. This of course will make you lose your data whenever you open another atlasmap instance. Remember currently AtlasMap is anyways a single user / single session application with no support for multi document editing functionality. So I don't see many chances to improve the situation. I could though add a Confirmation dialog which warns the user we are about to kill his old session. How about that?

**Comment by apupier** [comment]
> I could though add a Confirmation dialog which warns the user we are about to kill his old session. How about that?

this is the minimum but fine for a first iteration

**Comment by lhein** [comment]
I am open for your improvement suggestions as far as doable on our side of code.

**Comment by apupier** [comment]
now when calling several times "Open AtlasMap" command, it is stopping/starting AtlasMap and closing/reopeing the internal view (if internal view used). This is useless, no stop/start is needed as it is reloading the same content.

**Comment by lhein** [comment]
that is not true. We hand over an adm file. That file could have been changed.

**Comment by apupier** [comment]
> that is not true. We hand over an adm file. That file could have been changed.

there is no adm file provided when using the Open command from the palette

**Comment by lhein** [comment]
ok, I did not know you are talking about the open without adm file. I can add an exception for that case.

File: src/test/command.start.test.ts [hunk 4]
```diff
@@ -79,35 +79,6 @@ testUtils.BROWSER_TYPES.forEach(function (browserConfig) {
 				});
 		});
 
-		it("Test Start Command invocation with running AtlasMap instance (DO NOT SPAWN MORE THAN ONE ATLASMAP)", function(done) {
```
Comment (apupier) [inline src/test/command.start.test.ts]: why removing this test? We still don't want to have a second AtlasMap instance to be started when calling Open twice.

File: src/extension.ts [hunk 5]
```diff
@@ -33,37 +58,65 @@ export function activate(context: vscode.ExtensionContext) {
...
 
-function launchAtlasMapLocally(atlasmapExecutablePath: string, port: string): Promise<any> {
+function isAtlasMapRunning(): boolean {
+	return atlasMapProcess !== undefined && atlasMapLaunchPort !== undefined;
+}
+
+async function ensureNoOtherAtlasMapInstanceRunning() {
+	if (isAtlasMapRunning()) {
+		// we need to stop a running atlasmap to make the next import work
+		let choice = await vscode.window.showWarningMessage("The is currently a local AtlasMap instance running. We need to restart that instance. Make sure you have saved all your changes in the AtlasMap UI to prevent data loss.", { modal: true }, "Restart");
```
Comment (apupier) [inline src/extension.ts]: typo "The is" --> "There is"

Comment (lhein) [reply src/extension.ts]: fixed

File: src/test/command.start.test.ts (hunk 4 above)
Comment (lhein) [reply src/test/command.start.test.ts]: added it back

File: src/extension.ts [hunk 6]
```diff
@@ -12,58 +12,119 @@ let atlasMapServerOutputChannel: vscode.OutputChannel;
...
+								admFilePath = localAdmFile;
+							})					
+							.catch( (err) => {
+								vscode.window.showErrorMessage("Unable to start AtlasMap instance");
+								console.error(err);
+							});
+					})
+					.catch( (err) => {
+						vscode.window.showErrorMessage("Unable to start AtlasMap instance");
+						console.error(err);
```
Comment (apupier) [inline src/extension.ts]: I think we need to use a specific outputchannel, otherwise normal users won't see the error logged in the console

Comment (lhein) [reply src/extension.ts]: this is an exception. Those are usually output to the console.error stream.

**Comment by lhein** [comment]
> I'm wondering if on the "Restart" button, it shouldn't be more precise like: "Restart AtlasMap" or "Restart AtlasMap to import Data Transformation" to avoid that users thinks that it is the whole VS Code instance that will restart.

The text in the warning dialog starts with "There is currently a local AtlasMap instance running. We need to restart that instance." I think it is pretty much clear what we intend to restart here?

> seems that you are not planning to write a test o,n this PR, if true, can you create the corresponding ticket? it is a really important use case fro which not having regression is important.

Which use case do u mean?

> did manual tests, works great, not detected issues

Great. So it works on both Linux and Windows.

**Comment by apupier** [comment]
> > seems that you are not planning to write a test on this PR, if true, can you create the corresponding ticket? it is a really important use case fro which not having regression is important.
> 
> Which use case do u mean?

opening an adm file in AtlasMap UI
which can be subdivised in several cases:
- opening when there is no AtlasMap running
- opening when there is an already opened AtlasMap instance
- opening when a file has an .adm extension but is not an AtlasMap file

**Comment by apupier** [comment]
> > I'm wondering if on the "Restart" button, it shouldn't be more precise like: "Restart AtlasMap" or "Restart AtlasMap to import Data Transformation" to avoid that users thinks that it is the whole VS Code instance that will restart.
> 
> The text in the warning dialog starts with "There is currently a local AtlasMap instance running. We need to restart that instance." I think it is pretty much clear what we intend to restart here?

The first time I noticed it, i thought it was restarting the whole VS Code instance but it might be clear enough for most users.

File: src/extension.ts (hunk 6 above)
Comment (apupier) [reply src/extension.ts]: i think that VS Code is recommending to use outputchannels per extension https://code.visualstudio.com/api/extension-capabilities/common-capabilities#output-channel
it allows to differentiate general errors to errors specific to an extension.
If you can consider that it is too much work, another task can be created to tackle it.

Comment (lhein) [reply src/extension.ts]: created https://github.com/jboss-fuse/vscode-atlasmap/issues/97 to work on that

**Comment by lhein** [comment]
> opening when a file has an .adm extension but is not an AtlasMap file

What is the value of this test? It doesn't really matter from the VSCode side of things. We just hand it over.

**Comment by apupier** [comment]
> > opening when a file has an .adm extension but is not an AtlasMap file
> 
> What is the value of this test? It doesn't really matter from the VSCode side of things. We just hand it over.

is it handled well on AtlasMap side? or is it exploding? Can we recover from that? Are we losing data?
depending on the choice (and the evolution of this choice) on how to handle this case in AtlasMap, we might need to check for it on our side, before giving it to AtlasMap, or not but we still need to have a test to check that it doesn't change.

**Comment by lhein** [comment]
added the 3 test cases

File: src/test/command.test.utils.ts [hunk 7]
```diff
@@ -9,6 +9,9 @@ import * as vscode from "vscode";
 import AtlasMapPanel from '../atlasMapWebView';
 
 const request = require("request");
+const fileUrl = require('file-url');
```
Comment (apupier) [inline src/test/command.test.utils.ts]: use of import * as from instead of require is the new way to do it.
SonarLint is making a soft recommendations to change

image

File: src/test/command.test.utils.ts [hunk 8]
```diff
@@ -156,4 +160,24 @@ export function createExecuteCommandStubFakingExternalOpenBrowserCall() {
 		});
 	executeCommandStub.callThrough();
 	return executeCommandStub;
-}
\ No newline at end of file
+}
+
+export function downloadTestADM() {
+	let f = "./mockdocfhir.adm";
+	let url: string = "https://github.com/atlasmap/atlasmap/raw/master/ui/test-resources/adm/mockdocfhir.adm";
```
Comment (apupier) [inline src/test/command.test.utils.ts]: I'm wondering if we should not point to a tag version, the one correspondign to the jar that we are embedding.
it will avoid that out tests are broken in case the master if atlasmap is refactored or if the adm file is no more compatible.

File: src/test/command.start.test.ts [hunk 9]
```diff
@@ -9,18 +9,24 @@ import * as testUtils from "./command.test.utils";
 import * as vscode from "vscode";
 import { BrowserType } from "../utils";
 
+const fs = require("fs");
 const expect = chai.expect;
 chai.use(sinonChai);
 
+const WARN_MSG:string = "There is currently a local AtlasMap instance running. We need to restart that instance. Make sure you have saved all your changes in the AtlasMap UI to prevent data loss.";
```
[the remark this answers is not in the document]
Comment (lhein) [reply src/test/command.start.test.ts]: done

File: src/test/command.test.utils.ts (hunk 7 above)
Comment (lhein) [reply src/test/command.test.utils.ts]: done

File: src/test/command.test.utils.ts (hunk 8 above)
Comment (lhein) [reply src/test/command.test.utils.ts]: done

**Comment by lhein** [comment]
the atlasmap version is now maintained in the package.json file and both the pre-install script and the test-utils script use it from there.

File: package.json [hunk 10]
```diff
@@ -4,6 +4,9 @@
 	"version": "0.0.1",
 	"publisher": "jboss-fuse",
 	"preview": true,
+	"config": {
+		"atlasmapversion": "1.39.4"
```
Comment (apupier) [inline package.json]: didn't know how that way to define global properties, it is cool
